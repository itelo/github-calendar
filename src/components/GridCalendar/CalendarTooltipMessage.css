.calendar-tooltip-container {
  color: #959da5;
  margin: 0;
  font-size: 13px;
  padding: 40;
}

.bold-text {
  font-weight: bold;
  color: #dfe2e5;
}
