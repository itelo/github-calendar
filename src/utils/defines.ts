export const CONTRIBUTIONS_MOST = 35;
export const CONTRIBUTIONS_ALOT = 28;
export const CONTRIBUTIONS_MODERATE = 21;
export const CONTRIBUTIONS_LITTLE = 14;

export const MAX_DAYS_IN_YEAR = 366;
export const MAX_DAYS_IN_WEEK = 7;

export const TOTAL_WEEK_DAYS = 7;
export const TOTAL_MONTH_COUNT = 12;
