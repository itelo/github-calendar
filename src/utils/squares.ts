import {
  CONTRIBUTIONS_MOST,
  CONTRIBUTIONS_ALOT,
  CONTRIBUTIONS_MODERATE,
  CONTRIBUTIONS_LITTLE
} from "./defines";

export const getColorClass = (count: number) => {
  switch (true) {
    case count >= CONTRIBUTIONS_MOST:
      return "most";
    case count >= CONTRIBUTIONS_ALOT:
      return "alot";
    case count >= CONTRIBUTIONS_MODERATE:
      return "moderate";
    case count >= CONTRIBUTIONS_LITTLE:
      return "little";
    default:
      return "nothing";
  }
};
