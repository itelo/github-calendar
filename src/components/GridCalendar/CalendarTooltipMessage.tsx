import * as React from "react";
import dfns from "date-fns";
import "./CalendarTooltipMessage.css";

type CalendarTooltipMessageProps = {
  date: Date;
  contribuitions: number;
};

const CalendarTooltipMessage: React.FunctionComponent<
  CalendarTooltipMessageProps
> = ({ date, contribuitions }) => {
  const formattedDate = `${dfns.format(date, "[on] MMM DD[,] YYYY")}`;

  switch (true) {
    case contribuitions === 0:
      return (
        <p
          data-testid="calendar-tooltip-container"
          className="calendar-tooltip-container"
        >
          <span data-testid="bold-text" className="bold-text">
            No contributions{" "}
          </span>
          {formattedDate}
        </p>
      );
    case contribuitions === 1:
      return (
        <p
          data-testid="calendar-tooltip-container"
          className="calendar-tooltip-container"
        >
          <span data-testid="bold-text" className="bold-text">
            1 contribution{" "}
          </span>
          {formattedDate}
        </p>
      );
    default:
      return (
        <p
          data-testid="calendar-tooltip-container"
          className="calendar-tooltip-container"
        >
          <span data-testid="bold-text" className="bold-text">
            {contribuitions} contributions{" "}
          </span>
          {formattedDate}
        </p>
      );
  }
};

export default CalendarTooltipMessage;
