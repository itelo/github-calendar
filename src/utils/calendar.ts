import dfns from "date-fns";
import * as R from "ramda";
import { MAX_DAYS_IN_YEAR } from "./defines";

export const getAllDaysBasedInAYear = (year: number) => {
  const today = new Date();
  const yearHelp = dfns.setYear(today, year);
  const maxDateBasedOnYearPassed = dfns.endOfYear(yearHelp);
  const lastDay = dfns.isAfter(maxDateBasedOnYearPassed, today)
    ? today
    : maxDateBasedOnYearPassed;

  const possibleFirstDay = dfns.subDays(lastDay, MAX_DAYS_IN_YEAR);

  const firstDay = findSaturday(possibleFirstDay);

  const rangeDays = dfns.differenceInCalendarDays(lastDay, firstDay);
  const rangeDaysToSubtract = R.reverse(R.range(0, rangeDays));

  return R.map(subtract(lastDay), rangeDaysToSubtract);
};

export const findSaturday = (date: Date) => {
  let count = 0;
  for (;;) {
    if (!dfns.isSaturday(dfns.subDays(date, count))) {
      count++;
    } else {
      break;
    }
  }
  return dfns.subDays(date, count);
};

export const subtract = (date: Date) => (day: number) =>
  dfns.subDays(date, day);
type TreeDays = {
  date: string;
  count: number;
};

export const concat = (treeDays: TreeDays[]) => (day: Date) => {
  const data = R.find(isDayInTreeDays(day), treeDays);
  if (data) {
    return {
      date: day,
      count: data.count
    };
  } else {
    return {
      date: day,
      count: 0
    };
  }
};

export const isDayInTreeDays = (day: Date) => ({ date }: { date: string }) =>
  dfns.isSameDay(new Date(date), day);

export const mergeTreeJsonWithAllDays = (
  allDays: Date[],
  treeDays: TreeDays[]
) => R.map(concat(treeDays), allDays);

export const extractMonthPositionFromData = (
  acc: { month: string; position: number; lastMonthHelper: string }[],
  data: {
    count: number;
    date: Date;
  }[],
  index: number
) => {
  const month = dfns.format(data[0].date, "MMM");

  if (acc.length === 0 || acc[acc.length - 1].lastMonthHelper !== month) {
    acc.push({
      month,
      position: index,
      lastMonthHelper: month
    });
  }
  return acc;
};
