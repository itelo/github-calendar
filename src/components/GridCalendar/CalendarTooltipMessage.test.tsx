import React from "react";
import { render } from "react-testing-library";
import "react-testing-library/cleanup-after-each";
import CalendarTooltipMessage from "./CalendarTooltipMessage";
import "jest-dom/extend-expect";
import { CONTRIBUTIONS_MOST } from "../../utils/defines";

describe("<CalendarTooltipMessage />", () => {
  test("test CalendarTooltipMessage passing mocked date and contribuitions=CONTRIBUTIONS_MOST", () => {
    const date = new Date("2016-10-22");
    const contribuitions = CONTRIBUTIONS_MOST;
    const { getByTestId, container } = render(
      <CalendarTooltipMessage date={date} contribuitions={contribuitions} />
    );

    const calendarTooltipcontainer = getByTestId(
      "calendar-tooltip-container"
    );
    expect(calendarTooltipcontainer).toHaveTextContent(
      `${CONTRIBUTIONS_MOST} contributions on Oct 21, 2016`
    );

    expect(container).toMatchSnapshot();
  });

  test("test CalendarTooltipMessage passing mocked date and contribuitions=1", () => {
    const date = new Date("2016-10-22");
    const contribuitions = 1;
    const { getByTestId, container } = render(
      <CalendarTooltipMessage date={date} contribuitions={contribuitions} />
    );

    const calendarTooltipcontainer = getByTestId(
      "calendar-tooltip-container"
    );
    expect(calendarTooltipcontainer).toHaveTextContent(
      `1 contribution on Oct 21, 2016`
    );

    expect(container).toMatchSnapshot();
  });

  test("test CalendarTooltipMessage passing mocked date and contribuitions=0", () => {
    const date = new Date("2016-10-22");
    const contribuitions = 0;
    const { getByTestId, container } = render(
      <CalendarTooltipMessage date={date} contribuitions={contribuitions} />
    );

    const calendarTooltipcontainer = getByTestId(
      "calendar-tooltip-container"
    );
    expect(calendarTooltipcontainer).toHaveTextContent(
      `No contributions on Oct 21, 2016`
    );

    expect(container).toMatchSnapshot();
  });
});
