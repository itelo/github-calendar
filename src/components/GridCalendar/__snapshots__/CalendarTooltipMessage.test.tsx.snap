// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarTooltipMessage /> test CalendarTooltipMessage passing mocked date and contribuitions=0 1`] = `
<div>
  <p
    class="calendar-tooltip-container"
    data-testid="calendar-tooltip-container"
  >
    <span
      class="bold-text"
      data-testid="bold-text"
    >
      No contributions
       
    </span>
    on Oct 21, 2016
  </p>
</div>
`;

exports[`<CalendarTooltipMessage /> test CalendarTooltipMessage passing mocked date and contribuitions=1 1`] = `
<div>
  <p
    class="calendar-tooltip-container"
    data-testid="calendar-tooltip-container"
  >
    <span
      class="bold-text"
      data-testid="bold-text"
    >
      1 contribution
       
    </span>
    on Oct 21, 2016
  </p>
</div>
`;

exports[`<CalendarTooltipMessage /> test CalendarTooltipMessage passing mocked date and contribuitions=CONTRIBUTIONS_MOST 1`] = `
<div>
  <p
    class="calendar-tooltip-container"
    data-testid="calendar-tooltip-container"
  >
    <span
      class="bold-text"
      data-testid="bold-text"
    >
      35
       contributions
       
    </span>
    on Oct 21, 2016
  </p>
</div>
`;
